### PR TITLE
added comments to sale, ref #262

### PR DESCRIPTION
**Pull request by maxhansson**
wip

### CODE REVIEW

**Comment by maxhansson** [comment]
Worth noting

The comment's location on the invoice
Only the first comment is posted on the invoice
the returnpath after adding a comment is edit_

File: app/models/sale.rb [hunk 1]
```diff
@@ -359,4 +360,8 @@ def has_document?
   def default_values
     self.payment_term = DEFAULT_PAYMENT_TERM
   end
+
+  def parent_name
+    '#'+id.to_s
```
Comment (qzio) [inline app/models/sale.rb]: Shouldn't this be the `invoice_number`?
We never show the id from the database to the user anywhere else...

File: app/controllers/sales_controller.rb [hunk 2]
```diff
@@ -38,7 +38,13 @@ def new
     @sale.customer_id = params[:customer_id] if params[:customer_id]
   end
 
+  # returnpath from adding comments
+  def edit
+    redirect_to sale_path(@sale)
+  end
+
```
Comment (qzio) [inline app/controllers/sales_controller.rb]: It must be possible to direct to the `#show` method directly without going to `#edit` first.

File: app/views/sales/_invoice.pdf.haml [hunk 3]
```diff
@@ -1,5 +1,8 @@
 - customer = sale.customer.decorate
 - organization = sale.organization.decorate
+- comment = ''
+- if sale.comments.first
+  - comment = sale.comments.first.body
```
Comment (qzio) [inline app/views/sales/_invoice.pdf.haml]: I'm not sure I think this is a good idea.
I think I would rather add a "comment" database field.

However, sales should still have comments so this PR is still valid.

Comment (maxhansson) [inline app/views/sales/_invoice.pdf.haml]: You have a point of invoice comment as a database field. The difference between text on invoice and internal comments becomes clearer to the user. The text is set when the sale-head is registered. 

Comments are saved in one place but return to the parent edit function. It matters in issue 276.

**Comment by maxhansson** [comment]
When will add / change invoice text be possible?

File: db/schema.rb [hunk 4]
```diff
@@ -201,11 +201,11 @@
     t.string   "name"
     t.string   "file_name"
     t.string   "run_type"
-    t.string   "state"
     t.integer  "accounting_plan_id"
     t.integer  "organization_id"
     t.datetime "created_at"
     t.datetime "updated_at"
+    t.string   "state"
```
Comment (qzio) [inline db/schema.rb]: I would prefer if we did not commit these kind of changes.
If it is ok by you, I can "revert" these db/schema.rb changes manually.

File: db/schema.rb [hunk 5]
```diff
@@ -278,9 +278,9 @@
     t.integer  "user_id"
     t.datetime "created_at"
     t.datetime "updated_at"
-    t.string   "state"
-    t.string   "download_file_name"
-    t.string   "download_content_type"
+    t.string   "state",                 limit: 255
```
Comment (qzio) [inline db/schema.rb]: I don't think this change is needed.

File: app/controllers/sales_controller.rb [hunk 6]
```diff
@@ -39,6 +39,7 @@ def new
   end
 
   def show
+    @object = @sale
```
Comment (qzio) [inline app/controllers/sales_controller.rb]: I don't think it's a good idea to have a variable named `@object`

Comment (maxhansson) [inline app/controllers/sales_controller.rb]: @sale.decorator is not working when adding comments

Comment (qzio) [inline app/controllers/sales_controller.rb]: that makes no sense... the @object here should not be needed... if the sale object without `.decorated` is actually needed, the name of the variable must reflect that, "object" could be anything.

suggestion: `@non_decorated_sale`

**Pull request merged**